Docker daemon panic when external registry auth service returns 401

Output of `docker version`:

```
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:37:01 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:37:01 2016
 OS/Arch:      linux/amd64
```


Output of `docker info`:

```
Containers: 66
 Running: 0
 Paused: 0
 Stopped: 66
Images: 63
Server Version: 1.10.2
Storage Driver: btrfs
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 3.16.0-36-generic
Operating System: Ubuntu 14.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.955 GiB
Name: <edited>
ID: C7NK:RNIQ:BYLM:I52O:EZAK:O7JP:XFZ2:7EPS:SZ4S:3GVO:E6YK:DZWO
```

Provide additional environment details (AWS, VirtualBox, physical, etc.):

**VirtualBox VM**

List the steps to reproduce the issue:
1. Setup distribution/docker version 2.3 with token authentication
2. Start a push/pull
3. When the docker daemon tries to get the authentication token, if the service returns the following response:

HTTP 401
ContentType: application/json

```json
{"error": "unauthorized"}
```

Describe the results you received:

**Docker daemon crashes**

Describe the results you expected:

It should log an error saying what is wrong with th response

Provide additional info you think is important:

Panic stack:

```
panic: runtime error: index out of range

goroutine 176 [running]:
github.com/docker/docker/distribution.retryOnError(0x7f44f1ad06b8, 0xc820d65b40, 0x0, 0x0)
        /usr/src/docker/.gopath/src/github.com/docker/docker/distribution/registry.go:143 +0x138
github.com/docker/docker/distribution.retryOnError(0x7f44f1ad06e0, 0xc820d41bc0, 0x0, 0x0)
        /usr/src/docker/.gopath/src/github.com/docker/docker/distribution/registry.go:150 +0x530
github.com/docker/docker/distribution.(*v2PushDescriptor).Upload(0xc820904f00, 0x7f44f1266420, 0xc820905280, 0x7f44f1266460, 0xc820a381e0, 0x0, 0x0)
        /usr/src/docker/.gopath/src/github.com/docker/docker/distribution/push_v2.go:335 +0x12fb
github.com/docker/docker/distribution/xfer.(*LayerUploadManager).makeUploadFunc.func1.1(0xc820a381e0, 0xc820a38060, 0x7f44f1266810, 0xc820904f00, 0xc820a9f7a0)
        /usr/src/docker/.gopath/src/github.com/docker/docker/distribution/xfer/upload.go:108 +0x1a6
created by github.com/docker/docker/distribution/xfer.(*LayerUploadManager).makeUploadFunc.func1
        /usr/src/docker/.gopath/src/github.com/docker/docker/distribution/xfer/upload.go:150 +0x13e
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,7 +16,8 @@ be found.
 
 ### Distribution
 
-- Fix a crash when pushing multiple images sharing the same layers to the same repository in parallel  [#20831](https://github.com/docker/docker/pull/20831)
+- Fix a crash when pushing multiple images sharing the same layers to the same repository in parallel [#20831](https://github.com/docker/docker/pull/20831)
+- Fix a panic when pushing images to a registry which uses a misconfigured token service [#21030](https://github.com/docker/docker/pull/21030)
 
 ### Plugin system
 

--- a/distribution/registry.go
+++ b/distribution/registry.go
@@ -140,7 +140,9 @@ func (th *existingTokenHandler) AuthorizeRequest(req *http.Request, params map[s
 func retryOnError(err error) error {
 	switch v := err.(type) {
 	case errcode.Errors:
-		return retryOnError(v[0])
+		if len(v) != 0 {
+			return retryOnError(v[0])
+		}
 	case errcode.Error:
 		switch v.Code {
 		case errcode.ErrorCodeUnauthorized, errcode.ErrorCodeUnsupported, errcode.ErrorCodeDenied:

